selfie.c

```diff
@@ -4090,13 +4119,82 @@ void execute() {
     }
 }
 
+// memory that holds all information about a context except registers
+int* contexts;
+
+// memory that holds registers of each single process
+int* registerstable;
+
+
+void saveContext(process) {
+	*(contexts + process * 5 + 0) = pc;
+	*(contexts + process * 5 + 2) = reg_hi;
+	*(contexts + process * 5 + 3) = reg_lo;
+	*(contexts + process * 5 + 4) = ir;
+
+	// registers dont have to be saved as they're already at correct position
+}
+
+void loadContext(process) {
+	pc = 		*(contexts + process * 5 + 0); 
+	reg_hi =	*(contexts + process * 5 + 2);
+	reg_lo = 	*(contexts + process * 5 + 3);
+	ir =		*(contexts + process * 5 + 4);
+
+	// redirect pointer to process' registers
+ 	registers = (registers_all + process * 32 );
+}
+
+
 void run() {
+
+    // current instruction within process, e.g. instr 2/3 in process 2
+    int m;
+
+    // current process id (0..2 for NUM_BINARIES = 3)
+    int process;
+
+    // loop variables
+    int j=0;
+    int i=0;
+
+    
+    contexts = malloc(5 * 4 * NUM_BINARIES);
```

Comment:
usually, the number of concurrent processes is not a constant. consider changing that when it comes to forking